[Bug]: Preset resolution uses dirname() on package name instead of resolving path, breaks pnpm monorepos

## Describe the bug

Storybook's preset resolution in `core-server/load.ts` incorrectly uses `dirname()` on a package name string instead of resolving the package path first. This causes `ERR_MODULE_NOT_FOUND` errors in pnpm monorepos due to strict module resolution.

**The Bug:**
```javascript
// Current code (line 7641 in dist/core-server/index.js)
builderName && corePresets.push(join(dirname(builderName), "preset.js"))
```

When `builderName = "@storybook/builder-vite"`:
- `dirname("@storybook/builder-vite")` returns `"@storybook"` (string manipulation)
- Results in path: `"@storybook/preset.js"`
- pnpm's strict resolution cannot find this path

**Expected behavior:**
```javascript
// Should resolve the package first
builderName && corePresets.push(join(resolvePackageDir(builderName), "preset.js"))
```

This would:
- Resolve `"@storybook/builder-vite"` to actual filesystem path
- Results in: `/path/to/node_modules/@storybook/builder-vite/preset.js` ✅

## Reproduction link

[pnpm vue + storybook reproduction](https://github.com/braedenfoster/storybook-pnpm-repro)

## Reproduction steps

1. Create a pnpm workspace/monorepo
2. Install Storybook 10.2.x with `@storybook/addon-vitest`
3. Run tests with vitest `--project=storybook`
4. Observe the `ERR_MODULE_NOT_FOUND` error for `@storybook/preset.js`

Expected: Preset should load successfully
Actual: Error [ERR_MODULE_NOT_FOUND]: Cannot find module '@storybook/preset.js'

## System

```bash
Environment Info:
  Node: 24.11.0
  Package Manager: pnpm 10.22.0
  
  Storybook: 10.2.1
  @storybook/addon-vitest: 10.1.5
  @storybook/builder-vite: 10.2.4
  @storybook-vue/nuxt: 9.1.0-29411911.f34c865
  
  Framework: Nuxt 4.2.1
  Builder: Vite
```

## Additional context

### Why this affects pnpm specifically:
- npm/yarn use flat node_modules
- pnpm uses strict nested `.pnpm` structure where `dirname()` on a package name creates invalid paths

### Verified Impact:
- Bug exists in 10.2.1 through 10.2.4 (all stable releases)
- Bug exists in 10.3.0-alpha.2
- Bug exists in `next` branch source code (verified in `code/core/src/core-server/load.ts`)

### Proposed Fix:

**File:** `code/core/src/core-server/load.ts`

```diff
- builderName && corePresets.push(join(dirname(builderName), "preset.js"))
+ builderName && corePresets.push(join(resolvePackageDir(builderName), "preset.js"))
```

This uses Storybook's existing `resolvePackageDir()` utility (which is already used in the same file to resolve the "storybook" package). This is consistent with how `resolvedRenderer` is properly resolved via `resolveAddonName()`.

--- a/code/core/src/core-server/load.ts
+++ b/code/core/src/core-server/load.ts
@@ -11,7 +11,7 @@ import type { BuilderOptions, CLIOptions, LoadOptions, Options } from 'storybook
 
 import { global } from '@storybook/global';
 
-import { join, relative, resolve } from 'pathe';
+import { dirname, isAbsolute, join, relative, resolve } from 'pathe';
 
 import { resolvePackageDir } from '../shared/utils/module';
 
@@ -68,7 +68,13 @@ export async function loadStorybook(
   const builderName = typeof builder === 'string' ? builder : builder?.name;
 
   if (builderName) {
-    corePresets.push(join(resolvePackageDir(builderName), 'preset.js'));
+    /* builderName can be a bare package name (e.g. '@storybook/builder-vite') or an already-resolved
+       file URL / absolute path (e.g. 'file:///.../.../dist/index.js'). For bare package names, we
+       need to resolve the package directory first; for already-resolved paths, dirname works directly.
+    */
+    const isResolved = builderName.startsWith('file:') || isAbsolute(builderName);
+    const builderPresetDir = isResolved ? dirname(builderName) : resolvePackageDir(builderName);
+    corePresets.push(join(builderPresetDir, 'preset.js'));
   }
 
   // Load second pass: all presets are applied in order

--- a/code/core/src/core-server/load.ts
+++ b/code/core/src/core-server/load.ts
@@ -68,7 +68,7 @@ export async function loadStorybook(
   const builderName = typeof builder === 'string' ? builder : builder?.name;
 
   if (builderName) {
-    corePresets.push(join(dirname(builderName), 'preset.js'));
+    corePresets.push(join(resolvePackageDir(builderName), 'preset.js'));
   }
 
   // Load second pass: all presets are applied in order

--- a/code/core/src/core-server/load.ts
+++ b/code/core/src/core-server/load.ts
@@ -11,7 +11,7 @@ import type { BuilderOptions, CLIOptions, LoadOptions, Options } from 'storybook
 
 import { global } from '@storybook/global';
 
-import { dirname, join, relative, resolve } from 'pathe';
+import { join, relative, resolve } from 'pathe';
 
 import { resolvePackageDir } from '../shared/utils/module';
 

--- a/code/core/src/shared/utils/module.ts
+++ b/code/core/src/shared/utils/module.ts
@@ -37,8 +37,8 @@ export const resolvePackageDir = (
       return dirname(fileURLToPath(importMetaResolve(join(pkg, 'package.json'))));
     } catch {
       // Fallback using require.resolve for strict pnpm environments where import.meta.resolve may fail
-      const require = createRequire(import.meta.url);
-      return dirname(require.resolve(join(pkg, 'package.json')));
+      const req = createRequire(parent ?? import.meta.url);
+      return dirname(req.resolve(join(pkg, 'package.json')));
     }
   }
 };

--- a/code/core/src/shared/utils/module.ts
+++ b/code/core/src/shared/utils/module.ts
@@ -32,8 +32,14 @@ export const resolvePackageDir = (
   try {
     return dirname(fileURLToPath(importMetaResolve(join(pkg, 'package.json'), parent)));
   } catch {
-    // Necessary fallback for Bun runtime
-    return dirname(fileURLToPath(importMetaResolve(join(pkg, 'package.json'))));
+    try {
+      // Necessary fallback for Bun runtime
+      return dirname(fileURLToPath(importMetaResolve(join(pkg, 'package.json'))));
+    } catch {
+      // Fallback using require.resolve for strict pnpm environments where import.meta.resolve may fail
+      const require = createRequire(import.meta.url);
+      return dirname(require.resolve(join(pkg, 'package.json')));
+    }
   }
 };
 

